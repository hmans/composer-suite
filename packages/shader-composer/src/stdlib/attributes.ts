import { $ } from "../expressions"
import { GLSLType, Unit } from "../units"

export const Attribute = <T extends GLSLType>(type: T, name: string) =>
  Unit(type, $`${name}`, {
    name: `Attribute: ${name}`,
    varying: true,
    vertex: {
      header: $`attribute ${type} ${name};`
    }
  })
