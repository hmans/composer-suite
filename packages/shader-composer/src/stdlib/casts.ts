import { $ } from "../expressions"
import { type } from "../glslType"
import { GLSLType, Input, isUnit, Unit, UnitConfig } from "../units"
import { Float, Mat3, Mat4, Vec2, Vec3, Vec4 } from "./values"

export const float = (
  v: Input<"float" | "bool" | "int"> = 0,
  extras?: Partial<UnitConfig<"float">>
) => Float($`float(${v})`, extras)

export const vec2 = (
  x: Input<"float"> = 0,
  y: Input<"float"> = 0,
  extras?: Partial<UnitConfig<"vec2">>
) => Vec2($`vec2(${x}, ${y})`, extras)

export const vec3 = (
  x: Input<"float"> = 0,
  y: Input<"float"> = 0,
  z: Input<"float"> = 0,
  extras?: Partial<UnitConfig<"vec3">>
) => Vec3($`vec3(${x}, ${y}, ${z})`, extras)

export const vec4 = (
  x: Input<"float"> = 0,
  y: Input<"float"> = 0,
  z: Input<"float"> = 0,
  w: Input<"float"> = 0,
  extras?: Partial<UnitConfig<"vec4">>
) => Vec4($`vec4(${x}, ${y}, ${z}, ${w})`, extras)

/** Cast the given value to a mat3. */
export const mat3 = (i: Input<"mat3" | "mat4">) => Mat3($`mat3(${i})`)

/** Cast the given value to a mat4. */
export const mat4 = (i: Input<"mat3" | "mat4">) => Mat4($`mat4(${i})`)

export const unit = <T extends GLSLType>(i: Input<T>, config?: Partial<UnitConfig<T>>) =>
  isUnit(i) ? i : Unit(type(i), i, config)

/**
 * Wraps an input value into a unit that is configured to use a varying.
 *
 * @param i Input value (unit, JS value or expression) to wrap in a varying unit.
 * @param config Optional extra configuration for the newly created unit.
 * @returns A new unit that wraps the given value and is configured to use a varying.
 */
export const varying = <T extends GLSLType>(
  i: Input<T>,
  config?: Partial<UnitConfig<T>>
) => Unit(type(i), i, { ...config, varying: true })
