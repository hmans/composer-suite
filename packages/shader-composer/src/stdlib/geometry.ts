/** @packageDocumentation
 * Geometryyyy
 */

import { Vector2 } from "three"
import { $, Expression } from "../expressions"
import { injectAPI, GLSLType, Input, Unit } from "../units"
import { localToViewSpace, localToWorldSpace } from "./spaces"
import { Bool, Mat4, Vec2, Vec3 } from "./values"

export const CameraPosition = Vec3($`cameraPosition`, {
  name: "Camera Position"
})

export const ViewMatrix = Mat4($`viewMatrix`, {
  name: "View Matrix"
})

export const ModelMatrix = Mat4($`modelMatrix`, {
  name: "Model Matrix"
})

export const ModelViewMatrix = Mat4($`modelViewMatrix`, {
  name: "ModelView Matrix"
})

export const NormalMatrix = Mat4($`normalMatrix`, {
  name: "Normal Matrix"
})

export const ProjectionMatrix = Mat4($`projectionMatrix`, {
  name: "Projection Matrix"
})

/**
 * Returns true if instanced rendering is enabled, false if it is not.
 */
export const UsingInstancing = Bool($`
  #ifdef USE_INSTANCING
    true
  #else
    false
  #endif
`)

/**
 * Returns the instance matrix. Please note that this is only available when
 * instanced rendering is enabled.
 */
export const InstanceMatrix = Mat4($`instanceMatrix`, {
  name: "Instance Matrix",
  only: "vertex"
})

export const UV = Vec2($`uv`, {
  name: "UV",
  varying: true
})

const Vec3WithSpaceConversions = (expr: Expression, name: string) =>
  injectAPI(Vec3(expr, { name, varying: true }), (unit) => ({
    get world() {
      return Vec3(localToWorldSpace(unit), {
        varying: true,
        name: `${name} (World Space)`
      })
    },

    get view() {
      return Vec3(localToViewSpace(unit), {
        varying: true,
        name: `${name} (View Space)`
      })
    }
  }))

export const VertexPosition = Vec3WithSpaceConversions($`position`, "Vertex Position")
export const VertexNormal = Vec3WithSpaceConversions($`normal`, "Vertex Normal")

export const ViewDirection = Vec3(
  $`vec3(-${ViewMatrix}[0][2], -${ViewMatrix}[1][2], -${ViewMatrix}[2][2])`,
  { varying: true, name: "View Direction" }
)

export const IsFrontFacing = Bool($`gl_FrontFacing`, { only: "fragment" })

export const TilingUV = (
  uv: Input<"vec2"> = UV,
  tiling: Input<"vec2"> = new Vector2(1, 1),
  offset: Input<"vec2"> = new Vector2(0, 0)
) =>
  Vec2(
    $`vec2(
      ${uv}.x * ${tiling}.x + ${offset}.x,
      ${uv}.y * ${tiling}.y + ${offset}.y)`
  )

export const Attribute = <T extends GLSLType>(type: T, name: string) =>
  Unit(type, $`${name}`, {
    name: `Attribute: ${name}`,
    varying: true,
    vertex: {
      header: $`attribute ${type} ${name};`
    }
  })
