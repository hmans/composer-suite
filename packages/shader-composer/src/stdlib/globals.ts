import { $ } from "../expressions"
import { Int, Vec2 } from "./values"

/**
 * Returns the current fragment's on-screen coordinate.
 */
export const FragmentCoordinate = Vec2($`gl_FragCoord.xy`, {
  name: "Fragment Coordinate",
  only: "fragment"
})

/**
 * In instanced rendering, will return the instance ID.
 * Wraps the `gl_InstanceID` GLSL built-in.
 *
 * Note: available in vertex shader only!
 */
export const InstanceID = Int($`gl_InstanceID`, {
  name: "Instance ID",
  only: "vertex"
})

/**
 * Returns the verte ID.
 * Wraps the `gl_VertexID` GLSL built-in.
 *
 * Note: available in vertex shader only!
 */
export const VertexID = Int($`gl_VertexID`, {
  name: "Vertex ID",
  only: "vertex"
})
