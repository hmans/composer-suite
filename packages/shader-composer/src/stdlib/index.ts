/** @packageDocumentation
 * The `shader-composer/stdlib` module contains a core collection of Shader unit implementations.
 */

export * from "./artistic"
export * from "./casts"
export * from "./geometry"
export * from "./globals"
export * from "./logic"
export * from "./masters"
export * from "./math"
export * from "./rotation"
export * from "./scene"
export * from "./spaces"
export * from "./textures"
export * from "./uniforms"
export * from "./values"
export * from "./vectors"
