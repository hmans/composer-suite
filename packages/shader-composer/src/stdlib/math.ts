import { pipe } from "fp-ts/function"
import { $ } from "../expressions"
import { type } from "../glslType"
import { GLSLType, Input, Unit } from "../units"
import { Float } from "./values"

/**
 * @internal
 */
export const Operator =
  (name: string, operator: "+" | "-" | "*" | "/") =>
  <A extends GLSLType, B extends GLSLType>(a: Input<A>, b: Input<B>) =>
    Unit(type(a), $`${a} ${operator} ${b}`, {
      name: `${name} (${type(a)})`
    })

/**
 * @internal
 */
export const SingleArgumentFunction =
  <TA extends GLSLType = "float" | "vec2" | "vec3" | "vec4">(
    name: string,
    functionName: string
  ) =>
  <A extends TA>(a: Input<A>) =>
    Unit(type(a), $`${functionName}(${a})`, { name: `${name} (${type(a)})` })

/**
 * A Shader Unit that adds two values and returns the result.
 */
export const Add = Operator("Add", "+")

/**
 * A Shader Unit that subtracts two values and returns the result.
 */
export const Sub = Operator("Subtract", "-")

export const Subtract = Sub

/**
 * A Shader Unit that multiplies two values and returns the result.
 */
export const Mul = Operator("Multiply", "*")

export const Multiply = Mul

/**
 * A Shader Unit that divides two values and returns the result.
 */
export const Div = Operator("Divide", "/")

export const Divide = Div

/**
 * Calculates the sine value of the input value.
 */
export const Sin = SingleArgumentFunction("Sin", "sin")

/**
 * Calculates the cosine value of the input value.
 */
export const Cos = SingleArgumentFunction("Cos", "cos")

export const Tan = SingleArgumentFunction("Tan", "tan")

export const Asin = SingleArgumentFunction("Asin", "asin")

export const Acos = SingleArgumentFunction("Acos", "acos")

export const Pow = <T extends "float" | "vec2" | "vec3" | "vec4">(
  a: Input<T>,
  e: Input<T>
) => Unit(type(a), $`pow(${a}, ${e})`, { name: `Pow (${type(a)})` })

export const Exp = SingleArgumentFunction("Exp", "exp")

export const Exp2 = SingleArgumentFunction("Exp2", "exp2")

export const Log = SingleArgumentFunction("Log", "log")

export const Log2 = SingleArgumentFunction("Log2", "log2")

export const Sqrt = SingleArgumentFunction("Sqrt", "sqrt")

export const InverseSqrt = SingleArgumentFunction("InverseSqrt", "inversesqrt")

/**
 * A Shader Unit that finds the nearest integer less than or equal to the input value.
 * It is equivalent to the GLSL expression `trunc(a)`.
 *
 * @param a The input value.
 * @returns The truncated value of `a`.
 */
export const Trunc = SingleArgumentFunction("Trunc", "trunc")

/**
 * A Shader Unit that finds the nearest integer to the input value.
 * It performs the GLSL expression `round(a)`.
 *
 * @param a The input value.
 * @returns The rounded value of `a`.
 */
export const Round = SingleArgumentFunction("Round", "round")

export const Fract = SingleArgumentFunction("Fract", "fract")

export const Floor = SingleArgumentFunction("Floor", "floor")

export const Ceil = SingleArgumentFunction("Ceil", "ceil")

export const Abs = SingleArgumentFunction("Abs", "abs")

export const Sign = SingleArgumentFunction("Sign", "sign")

/**
 * Converts the given value from degrees to radians.
 */
export const Radians = SingleArgumentFunction("Radians", "radians")

/**
 * Converts the given value from radians to degrees.
 */
export const Degrees = SingleArgumentFunction("Degrees", "degrees")

export const Modulo = <
  A extends "float" | "vec2" | "vec3" | "vec4",
  B extends "float" | A
>(
  a: Input<A>,
  b: Input<B>
) => Unit(type(a), $`mod(${a}, ${b})`)

export const Clamp = <T extends GLSLType>(
  x: Input<T>,
  min: Input<T>,
  max: Input<T>
) => Unit(type(x), $`clamp(${x}, ${min}, ${max})`, { name: "Clamp" })

export const Clamp01 = (x: Input<"float">) => Clamp(x, 0, 1)

export const Saturate = Clamp01

export const OneMinus = (v: Input<"float">) =>
  Float(Sub(1, v), { name: "OneMinus" })

/**
 * Negates the value (equivalent to multiplying it with -1.)
 *
 * @param v Value to negate.
 * @returns A shader unit holding the negated value.
 */
export const Negate = <T extends GLSLType>(v: Input<T>) =>
  Unit(type(v), Mul(v, -1), { name: "Negate" })

/**
 * Lerpy fun!
 *
 * @group Math
 * @param a
 * @param b
 * @param ratio
 * @returns
 */
export const $lerp = <T extends GLSLType>(
  a: Input<T>,
  b: Input<T>,
  ratio: Input<"float">
) => $`mix(${a}, ${b}, ${ratio})`

/**
 * Performs linear interpolation between two values, and returns the result.
 *
 * Wraps the GLSL `mix` function.
 *
 * @param a The starting value of the interpolation.
 * @param b The ending value of the interpolation.
 * @param ratio The interpolation ratio.
 * @returns The interpolated value.
 */
export const Lerp = <T extends GLSLType>(
  a: Input<T>,
  b: Input<T>,
  ratio: Input<"float">
) => Unit(type(a), $lerp(a, b, ratio), { name: "Mix" })

export const $inverseLerp = <T extends GLSLType>(
  a: Input<T>,
  b: Input<T>,
  c: Input<T>
) => $`(${c} - ${a}) / (${b} - ${a})`

/**
 * Performs inverse linear interpolation between two values, and returns the result.
 * Given three values `a`, `b`, and `c`, the result is the ratio of `c` between `a` and `b`.
 * Also see {@link Lerp}.
 *
 * @param a The starting value of the interpolation.
 * @param b The ending value of the interpolation.
 * @param c The value to find the interpolation ratio of.
 * @returns The interpolation ratio of `c` between `a` and `b`.
 */
export const InverseLerp = <T extends GLSLType>(
  a: Input<T>,
  b: Input<T>,
  c: Input<T>
) => Unit(type(a), $inverseLerp(a, b, c), { name: "Inverse Lerp" })

export const Mix = Lerp

/**
 * Given an `edge` value and an input value `v`, returns 0 if the value is less than the
 * edge, and 1 if the value is greater than or equal to the edge.
 *
 * Wraps the GLSL `step` function.
 *
 * @example
 * Moooooooo
 *
 * ```jsx
 * Step(0.5, 0.25, f)
 * ```
 *
 * @param edge The edge value.
 * @param v The value to test.
 * @returns The result of the step function.
 */
export const Step = (edge: Input<"float">, v: Input<"float">) =>
  Float($`step(${edge}, ${v})`, { name: "Step" })

/**
 * Performs a Hermite interpolation between two values, and returns the result.
 *
 * Wraps the GLSL `smoothstep` function.
 *
 * @param min The lower edge of the Hermite einterpolation.
 * @param max The upper edge of the Hermite interpolation.
 * @param v The source value for the interpolation.
 * @returns The result of the Hermite interpolation.
 */
export const Smoothstep = (
  min: Input<"float">,
  max: Input<"float">,
  v: Input<"float">
) => Float($`smoothstep(${min}, ${max}, ${v})`, { name: "Smoothstep" })

export const $remap = <T extends "float" | "vec2" | "vec3" | "vec4">(
  value: Input<T>,
  inMin: Input<T>,
  inMax: Input<T>,
  outMin: Input<T>,
  outMax: Input<T>
) => $`mix(${outMin}, ${outMax}, ${$inverseLerp(inMin, inMax, value)})`

export const Remap = <T extends "float" | "vec2" | "vec3" | "vec4">(
  value: Input<T>,
  inMin: Input<T>,
  inMax: Input<T>,
  outMin: Input<T>,
  outMax: Input<T>
) => Unit(type(value), $remap(value, inMin, inMax, outMin, outMax))

export const NormalizePlusMinusOne = (f: Input<"float">) =>
  Remap(f, -1, 1, 0, 1)

export const Min = <T extends "float" | "vec2" | "vec3" | "vec4">(
  a: Input<T>,
  b: Input<T>
) => Unit(type(a), $`min(${a}, ${b})`)

export const Max = <T extends "float" | "vec2" | "vec3" | "vec4">(
  a: Input<T>,
  b: Input<T>
) => Unit(type(a), $`max(${a}, ${b})`)

/**
 * Applies scaling (multiplication) and offset (addition) to a given value.
 *
 * @param v The value to scale and offset.
 * @param scale The scale to apply. (Default: 1)
 * @param offset The offset to apply. (Default: 0)
 * @returns A unit holding the scaled and offset value.
 */
export const ScaleAndOffset = <T extends GLSLType>(
  v: Input<T>,
  scale: Input = 1,
  offset: Input = 0
) =>
  Unit(
    type(v),
    pipe(
      v,
      (v) => Mul(v, scale),
      (v) => Add(v, offset)
    ),
    { name: "Scale and Offset" }
  )
