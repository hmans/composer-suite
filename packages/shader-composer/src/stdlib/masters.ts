import { Color } from "three"
import { $ } from "../expressions"
import { Input } from "../units"
import { VertexPosition } from "./geometry"
import { Master } from "./values"

export type ShaderMaterialMasterProps = {
  color?: Input<"vec3">
  alpha?: Input<"float">
  position?: Input<"vec3">
}

export const ShaderMaterialMaster = ({
  color = new Color("red"),
  alpha = 1,
  position = VertexPosition
}: ShaderMaterialMasterProps = {}) =>
  Master({
    name: "ShaderMaterial Master",

    vertex: {
      body: $`
				gl_Position = projectionMatrix * modelViewMatrix * vec4(${position}, 1.0);
			`
    },

    fragment: {
      body: $`
				gl_FragColor = vec4(${color}, ${alpha});
			`
    }
  })

export type CustomShaderMaterialMasterProps = {
  position?: Input<"vec3">
  normal?: Input<"vec3">
  diffuseColor?: Input<"vec3">
  emissiveColor?: Input<"vec3">
  fragColor?: Input<"vec3">
  alpha?: Input<"float">
  roughness?: Input<"float">
  metalness?: Input<"float">
}

export const CustomShaderMaterialMaster = ({
  position,
  normal,
  diffuseColor,
  emissiveColor,
  fragColor,
  roughness,
  metalness,
  alpha
}: CustomShaderMaterialMasterProps = {}) =>
  Master({
    name: "CustomShaderMaterial Master",

    vertex: {
      body: $`
				${position !== undefined ? $`csm_Position.xyz = ${position};` : ""}
				${normal !== undefined ? $`csm_Normal = ${normal};` : ""}
			`
    },

    fragment: {
      body: $`
  			${alpha !== undefined ? $`csm_DiffuseColor.a = ${alpha};` : ""}
				${diffuseColor !== undefined ? $`csm_DiffuseColor.rgb = ${diffuseColor};` : ""}
				${emissiveColor !== undefined ? $`csm_Emissive = ${emissiveColor};` : ""}
				${fragColor !== undefined ? $`csm_FragColor = vec4(${fragColor}, ${alpha});` : ""}

        #if defined IS_MESHSTANDARDMATERIAL || defined IS_MESHPHYSICALMATERIAL
          ${roughness !== undefined ? $`csm_Roughness = ${roughness};` : ""}
          ${metalness !== undefined ? $`csm_Metalness = ${metalness};` : ""}
        #endif
			`
    }
  })
