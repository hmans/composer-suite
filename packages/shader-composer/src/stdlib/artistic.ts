import { pipe } from "../pipes"
import { GLSLType, Input, Unit } from "../units"
import { VertexNormal, ViewDirection } from "./geometry"
import { Abs, Add, Lerp, Mul, Pow, Saturate, Smoothstep } from "./math"
import { Float } from "./values"
import { Dot } from "./vectors"

export type FresnelProps = {
  normal?: Input<"vec3">
  alpha?: Input<"float">
  bias?: Input<"float">
  intensity?: Input<"float">
  power?: Input<"float">
  factor?: Input<"float">
}

export const Fresnel = ({
  normal = VertexNormal.world,
  bias = 0,
  intensity = 1,
  power = 2,
  factor = 1
}: FresnelProps = {}) =>
  Float(
    pipe(
      normal,
      (v) => Dot(ViewDirection, v),
      (v) => Add(factor, v),
      (v) => Abs(v),
      (v) => Pow(v, power),
      (v) => Mul(v, intensity),
      (v) => Add(v, bias),
      (v) => Saturate(v)
    ),
    { name: "Fresnel" }
  )

export type GradientStops<T extends GLSLType = "vec3"> = GradientStop<T>[]

export type GradientStop<T extends GLSLType = "vec3"> = [Input<T>, Input<"float">]

export const Gradient = <T extends GLSLType = "vec3">(
  f: Input<"float">,
  ...stops: GradientStops<T>
) => {
  let color = stops[0][0]

  for (let i = 1; i < stops.length; i++) {
    const stop = stops[i]
    const previous = stops[i - 1]
    color = Lerp(color, stop[0], Smoothstep(previous[1], stop[1], f))
  }

  return color as Unit<T>
}
