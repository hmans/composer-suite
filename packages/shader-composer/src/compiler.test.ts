import { compileShader } from "./compiler"
import { $, glsl } from "./expressions"
import { Snippet } from "./snippets"
import { Bool, Float, Master } from "./stdlib"

describe("compileShader", () => {
  it("compiles shader programs from the given unit", () => {
    const root = Bool(true)
    const [shader] = compileShader(root)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("embeds body chunks in scoped blocks, with a local `value` variable", () => {
    const root = Master({
      fragment: {
        body: $`gl_FragColor.rgb = vec3(1.0, 0.0, 0.0);`
      }
    })

    const [shader] = compileShader(root)
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("resolves dependencies to other units", () => {
    const f = Float(123)
    const root = Float(f)

    const [shader] = compileShader(root)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("resolves dependencies to other units from within expressions", () => {
    const f = Float(123)
    const root = Float(glsl`${f} * 2.0`)

    const [shader] = compileShader(root)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("resolves dependencies to expressions", () => {
    const a = glsl`123.0`
    const b = glsl`${a} + 4.0`
    const root = Float(glsl`${b}`)

    const [shader] = compileShader(root)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("allows resolves dependencies unique to specific programs", () => {
    const position = Float(1, { name: "Position (Vertex Only)" })
    const color = Float(2, { name: "Color (Fragment Only)" })

    const root = Master({
      vertex: { body: glsl`gl_Position = ${position};` },
      fragment: { body: glsl`gl_FragColor = ${color};` }
    })

    const [shader] = compileShader(root)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("supports snippets", () => {
    const double = Snippet(
      (name) => $`
			float ${name}(in float a) {
				return a * 2.0;
			}`
    )

    const root = Float($`${double}(1.0)`)

    const [shader] = compileShader(root)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  describe("when rendering units with registered uniforms", () => {
    const getShader = () => {
      const unitWithUniform = Float(0, {
        uniform: { value: 0 }
      })

      const [shader] = compileShader(unitWithUniform)

      return shader
    }

    it("adds uniform declarations to the program headers", () => {
      expect(getShader().vertexShader).toMatchSnapshot()
      expect(getShader().fragmentShader).toMatchSnapshot()
    })

    it("adds the uniform value object to the returned uniforms object", () => {
      expect(getShader().uniforms).toMatchInlineSnapshot(`
        {
          "u_Anonymous_1": {
            "value": 0,
          },
        }
      `)
    })
  })

  it("only includes the units needed in individual programs", () => {
    const a = Float(1, { name: "A" })
    const b = Float(2, { name: "B" })

    const master = Master({
      vertex: {
        body: $`foo = ${a};`
      },
      fragment: {
        body: $`bar = ${b}`
      }
    })

    const [shader] = compileShader(master)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("when encountering a varying unit, its value dependencies should only be resolved in the vertex program", () => {
    /* A unit that we expect only to show up in the vertex program. */
    const a = Float($`onlyInVertexProgramForSomeReason`, { name: "A" })

    /* A verying unit that sources `a`. It's also configured to use a varying, which
		means that its value expression will only appear in the vertex program -- for this reason,
		we also only need the value expression's dependencies in the vertex program. */
    const b = Float($`${a} + 1.0`, { name: "B", varying: true })

    /* A master unit that specifically uses the `b` unit from within the fragment program. */
    const root = Master({ name: "Master", fragment: { body: $`value = ${b}` } })

    const [shader] = compileShader(root)

    expect(shader.vertexShader).toMatchSnapshot()
    expect(shader.fragmentShader).toMatchSnapshot()
  })

  it("throws an error when encountering a unit that is not able to run in the requested program", () => {
    const a = Float(1, { name: "A", only: "vertex" })
    const root = Master({ fragment: { body: $`${a}` } })
    expect(() => compileShader(root)).toThrowErrorMatchingInlineSnapshot(
      `"Encountered a unit "A" that is only allowed in the vertex shader, but was encountered when compiling the fragment shader. Consider wrapping the value, or the derived value you're interested in, in a Unit that has a varying."`
    )
  })
})
