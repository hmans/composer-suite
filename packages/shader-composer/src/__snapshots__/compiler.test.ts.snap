// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileShader allows resolves dependencies unique to specific programs 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

void main()
{
  /*** UNIT: Position (Vertex Only) ***/
  float Position_Vertex_Only_1 = 1.0;
  
  
  /*** UNIT: Anonymous ***/
  bool Anonymous_3 = true;
  {
    bool value = Anonymous_3;
    gl_Position = Position_Vertex_Only_1;
    Anonymous_3 = value;
  }
  
  
}"
`;

exports[`compileShader allows resolves dependencies unique to specific programs 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

void main()
{
  /*** UNIT: Color (Fragment Only) ***/
  float Color_Fragment_Only_2 = 2.0;
  
  
  /*** UNIT: Anonymous ***/
  bool Anonymous_3 = true;
  {
    bool value = Anonymous_3;
    gl_FragColor = Color_Fragment_Only_2;
    Anonymous_3 = value;
  }
  
  
}"
`;

exports[`compileShader compiles shader programs from the given unit 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  bool Anonymous_1 = true;
  
  
}"
`;

exports[`compileShader compiles shader programs from the given unit 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  bool Anonymous_1 = true;
  
  
}"
`;

exports[`compileShader embeds body chunks in scoped blocks, with a local \`value\` variable 1`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  bool Anonymous_1 = true;
  {
    bool value = Anonymous_1;
    gl_FragColor.rgb = vec3(1.0, 0.0, 0.0);
    Anonymous_1 = value;
  }
  
  
}"
`;

exports[`compileShader only includes the units needed in individual programs 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

void main()
{
  /*** UNIT: A ***/
  float A_1 = 1.0;
  
  
  /*** UNIT: Anonymous ***/
  bool Anonymous_3 = true;
  {
    bool value = Anonymous_3;
    foo = A_1;
    Anonymous_3 = value;
  }
  
  
}"
`;

exports[`compileShader only includes the units needed in individual programs 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

void main()
{
  /*** UNIT: B ***/
  float B_2 = 2.0;
  
  
  /*** UNIT: Anonymous ***/
  bool Anonymous_3 = true;
  {
    bool value = Anonymous_3;
    bar = B_2
    Anonymous_3 = value;
  }
  
  
}"
`;

exports[`compileShader resolves dependencies to expressions 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0 + 4.0;
  
  
}"
`;

exports[`compileShader resolves dependencies to expressions 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0 + 4.0;
  
  
}"
`;

exports[`compileShader resolves dependencies to other units 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0;
  
  
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = Anonymous_1;
  
  
}"
`;

exports[`compileShader resolves dependencies to other units 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0;
  
  
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = Anonymous_1;
  
  
}"
`;

exports[`compileShader resolves dependencies to other units from within expressions 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0;
  
  
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = Anonymous_1 * 2.0;
  
  
}"
`;

exports[`compileShader resolves dependencies to other units from within expressions 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_1 = 123.0;
  
  
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = Anonymous_1 * 2.0;
  
  
}"
`;

exports[`compileShader supports snippets 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;


			float snippet_1(in float a) {
				return a * 2.0;
			}
void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = snippet_1(1.0);
  
  
}"
`;

exports[`compileShader supports snippets 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;


			float snippet_1(in float a) {
				return a * 2.0;
			}
void main()
{
  /*** UNIT: Anonymous ***/
  float Anonymous_2 = snippet_1(1.0);
  
  
}"
`;

exports[`compileShader when encountering a varying unit, its value dependencies should only be resolved in the vertex program 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

/*** UNIT: B ***/
varying float v_B_2;


void main()
{
  /*** UNIT: A ***/
  float A_1 = onlyInVertexProgramForSomeReason;
  
  
  /*** UNIT: B ***/
  float B_2 = A_1 + 1.0;
  v_B_2 = B_2;
  
  
  /*** UNIT: Master ***/
  bool Master_3 = true;
  
  
}"
`;

exports[`compileShader when encountering a varying unit, its value dependencies should only be resolved in the vertex program 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

/*** UNIT: B ***/
varying float v_B_2;


void main()
{
  /*** UNIT: B ***/
  float B_2 = v_B_2;
  
  
  /*** UNIT: Master ***/
  bool Master_3 = true;
  {
    bool value = Master_3;
    value = B_2
    Master_3 = value;
  }
  
  
}"
`;

exports[`compileShader when rendering units with registered uniforms adds uniform declarations to the program headers 1`] = `
"/*** PROGRAM: VERTEX ***/

precision highp float;

/*** UNIT: Anonymous ***/
uniform float u_Anonymous_1;


void main()
{
  /*** UNIT: Anonymous ***/
  
  
}"
`;

exports[`compileShader when rendering units with registered uniforms adds uniform declarations to the program headers 2`] = `
"/*** PROGRAM: FRAGMENT ***/

precision highp float;

/*** UNIT: Anonymous ***/
uniform float u_Anonymous_1;


void main()
{
  /*** UNIT: Anonymous ***/
  
  
}"
`;
